lib/elixir/lib/enum.ex

```diff
@@ -717,34 +720,76 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
-  def fetch(first..last, index) when is_integer(index),
-    do: fetch_range(first, last, index)
 
-  def fetch(enumerable, index)
-      when is_list(enumerable) and is_integer(index) and index >= 0 do
-    do_fetch(enumerable, index)
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    case enumerable do
+      first..last ->
+        fetch_range(first, last, index)
+
+      _ ->
+        case Enumerable.count(enumerable) do
+          {:ok, count} when (count + index) < 0 ->
+            :error
+
+          {:ok, count} ->
+            do_fetch(enumerable, count + index, Enumerable)
+
+          {:error, _} ->
+            enumerable
+            |> reverse
+            |> fetch_list((-index) - 1)
+        end
+    end
   end
 
-  def fetch(enumerable, index) when is_integer(index) and index >= 0 do
-    res =
-      Enumerable.reduce(enumerable, {:cont, 0}, fn(entry, acc) ->
-        if acc == index do
-          {:halt, entry}
-        else
-          {:cont, acc + 1}
+  def fetch(enumerable, index) when is_list(enumerable) and is_integer(index) do
+    fetch_list(enumerable, index)
+  end
+
+  def fetch(first..last, index) when is_integer(index) do
+    fetch_range(first, last, index)
+  end
+
+  def fetch(enumerable, index) when is_integer(index) do
+    case enumerable do
+      first..last ->
```

Comment:
sorry case shouldn't have been there.
the reason is that I don't want to patter match against ranges, before i pattern match on a list.